### ISSUE TITLE
Audit validation tags for configuration properties

### ISSUE THREAD

**Issue by bluekeyes** [gomods/athens#1491]
**Is your feature request related to a problem? Please describe.**

For some properties (I hit it on `DownloadMode`, but may apply to others), the zero value is invalid and the property is effectively required, but is not marked as required for validation.

**Describe the solution you'd like**

Audit the configuration type definitions to make sure the validation tags are correct for properties. I think the `required` tag is the most relevant, since a missing required property gives a clear error at startup, while the zero value for the type may give a confusing error later on.

**Describe alternatives you've considered**

Introduce another method to validate configuration. This seems non-ideal, since validation tags are already used for some properties.

**Additional context**

This issue is split out of #1336, which has some additional information.

**Comment by arschles** [comment]
@bluekeyes We should certainly get better at validating configurations. Since there are a lot of options, it's easy to forget something and make a mistake, so I think we should look holistically at the configuration. For example, if `DownloadMode` is set to `"redirect"`, then `DownloadURL` needs to be set. We should check that and report it and disallow Athens from starting.

What do you think?

**Comment by bluekeyes** [comment]
Yeah, it makes sense to me that the application would need a way to validate relationships between properties and that that would provide a better user experience than just saying if an option is required or not.

From a pure code perspective, I personally find it a bit confusing when there are multiple validation methods at play in a project, so long term, I think it could be valuable to decide on when to use the `validator.v9` library (if at all) vs. when to use any new method introduced here. But I'm also not very familiar with the Athens code base or development style.

**Comment by arschles** [comment]
@bluekeyes I don't have any experience with validator.v9, but it looks nice. The idea I have in mind is to centralize all validation of configuration in one place in the code, regardless of where the configuration values came from - file, environment variables, or anywhere else we define later on.

Ideally, the validations would come with good error messages too, when Athens is misconfigured. I'll check out validator.v9 to see if I can get it working in https://github.com/gomods/athens/pull/1492 too - thanks!
